### DISCUSSION TITLE
멀티플레이어 환경에서 텔레포트

### DISCUSSION THREAD

**Discussion by yoon-hw22** [naverz/zepeto-studio-kor#1090] [5. Q&A]
멀티플레이어 환경에서 텔레포트를 할 때 다른 유저 눈에는 뛰어가는것처럼 보이는데 이부분 바로 동기화 할 수 있는 방법 없나요?

**Comment by smlee2952** [comment]
텔레포트 시에 

1. 클라이언트에서 서버로 텔레포트 위치 정보를 전송
2. 서버에서는 broadcast를 통해 다른 클라이언트에게 전달받은 정보를 다시 전송
3. 다른 클라이언트들은 해당 메세지를 받아 해당 유저를 전달받은 위치로 텔레포트

해주면 됩니다.

1. 텔레포트 시 아래 코드로 서버에게 정보 전달

interface TeleportMessageModel {
    clientId: string,
    transform: UnityEngine.Transform
}

let message = { clientId : this.room.SessionId, transform : transform} as TeleportMessageModel
this.room.Send("Teleport", message);

2. 서버는 메세지를 전달해준 클라이언트를 제외하고 나머지 클라이언트들에게 broadcast로 메세지 전달

this.onMessage("Teleport", (client, message) => {
            this.broadcast("Teleport", message, { except: client});
});

3. 클라이언트들은 해당 메세지 정보로 특정 유저를 텔레포트

this.multiplay.RoomJoined += (room: Room) => {
            room.OnStateChange += this.OnStateChange;
            room.AddMessageHandler("Teleport", (message: TeleportMessageModel) => {
                ZepetoPlayers.instance.GetPlayer(message.clientId).character.Teleport();
            });
        };

이런 느낌 입니다.

**Reply by yoon-hw22** [reply]
client starter 의 Start() 함수에서 multiplay.RoomJoind 안에 3번을 넣고, 서버 역할 하는 index의 onJoin에 번을 넣고, client starter 에 텔레포트 할 때 호출 할 함수를 만들어서 1번을 넣으면 되는건가요? 순서가 이상한건지 1번의 session id를 못받아오네요 ..

**Reply by smlee2952** [reply]
순서는 맞는 것 같습니다. 
1번의 interface 정의는 client Starter 클래스 밖에서 해주시면 되고, this.room.SessionId는 서버가 켜져 있고 룸에 연결되어 있다면 있을텐데 이상하네요...

**Reply by yoon-hw22** [reply]
game object 로 client starter 받아와서 성공했습니다 감사합니다

**Reply by chiro-zp** [reply]
I can't get it to work.  Rewrote index.ts and clientstarterV2.ts. Should number 1 be written in Teleport.ts?
